core/commands/add.go

```diff
@@ -113,12 +113,16 @@ remains to be implemented.
 					return
 				}
 
-				err = n.Pinning.Pin(context.Background(), rootnd, true)
+				rnk, err := rootnd.Key()
 				if err != nil {
 					res.SetError(err, cmds.ErrNormal)
 					return
 				}
 
+				mp := n.Pinning.GetManual()
+				mp.RemovePinWithMode(rnk, pin.Indirect)
```

Comment:
On Wed, May 27, 2015 at 11:42:36AM -0700, Jeromy Johnson wrote:
> > is the worry that something will GC in-between, while adding? (if
> > so, ok makes sense).
> 
> This too, a poorly timed gc could end really badly for us. (although
> its still not perfectly safe)

I think the right approach here is to have transactional changes to
the pin tree (see also the commit sessions discussed in #1187).
Coupling that with something like Git's gc.pruneexpire to avoid
garbage-collecting recently-created objects (say, in the last week?)
would give us better safety here.  It would also protect folks from
accidentally GC'ing recent objects they'd forgotten to tag.